test/elixir_analyzer/exercise_test/feature/pipes_test.exs

```diff
@@ -0,0 +1,184 @@
+        def hi(name) do
+          foo(name)
+        end
+      end,
+      # This case matches, because feature looks down the AST
+      defmodule MyModule do
+        def hi(name) do
+          name |> foo(:ok)
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> foo
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> foo()
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> String.downcase() |> foo()
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> String.downcase() |> foo
+        end
+      end
+    ]
+  end
+
+  test_exercise_analysis "function not with one parameter",
+    comments_include: ["not one parameter"] do
+    [
+      defmodule MyModule do
+        def hi(name) do
+          foo()
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          foo
+        end
+      end
+    ]
+  end
+
+  test_exercise_analysis "function with three parameter",
+    comments_exclude: ["not three parameter"] do
+    [
+      defmodule MyModule do
+        def hi(name) do
+          foo(name, :ok, false)
+        end
+      end,
+      # This case matches, because feature looks down the AST
+      defmodule MyModule do
+        def hi(name) do
+          :woops |> foo(name, :ok, false)
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> foo(:ok, false)
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> String.downcase() |> foo(:ok, false)
+        end
+      end
+    ]
+  end
+
+  test_exercise_analysis "function not with three parameter",
+    comments_include: ["not three parameter"] do
+    [
+      defmodule MyModule do
+        def hi(name) do
+          foo(name, false)
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> foo(false)
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> String.downcase() |> foo
+        end
+      end
+    ]
+  end
+
+  test_exercise_analysis "function with one piped parameter",
+    comments_exclude: ["not one piped parameter", "not one piped parameter (no parens)"] do
+    [
+      defmodule MyModule do
+        def hi(name) do
+          name |> foo
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> foo()
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> String.downcase() |> foo()
+        end
+      end,
+      defmodule MyModule do
+        def hi(name) do
+          name |> String.downcase() |> foo
+        end
+      end
+    ]
+  end
+
+  test_exercise_analysis "function not with one piped parameter",
+    comments_include: ["not one piped parameter", "not one piped parameter (no parens)"] do
```

Comment:
This is question to make sure I understand the implementation: those two forms (with parens, without parens) will always be either both matched or both unmatched, right? Because we are adding the parens. So you cannot write a test that will pass this?
```elixir
  test_exercise_analysis "...",
    comments_exclude: ["not one piped parameter"],
    comments_include: ["not one piped parameter (no parens)"] do
```

Reply:
Exactly. 
They do exactly the same (but do not trigger the compiler error saying they are the same form, because the AST are a bit different).